design/envoy-json-logging-custom-fields-design.md

```diff
@@ -0,0 +1,146 @@
+   - response_duration=%RESPONSE_DURATION%
+```
+
+The following is the resulting JSON config for Envoy:
+
+```json
+{
+   "timestamp": "%START_TIME%",
+   "method:" "%REQ(:METHOD)%",
+   "content-id": "%REQ(X-CONTENT-ID)%",
+   "response_duration": "%RESPONSE_DURATION%"
+}
+```
+
+## Alternatives Considered
+
+### User-defined Translation Table
+As implemented in [#3033](https://github.com/projectcontour/contour/pull/3033), another approach is to allow users to specify an additional translation table.
+This mirrors current default translation table, but allows users to define custom translations.
+Validation could be done in two stages:
+  1. Validate that all the right-hand-sides of translations appear valid.
+  2. Validate that all entries in `json-fields` are in the default or user-defined translation tables.
+However, [#3033](https://github.com/projectcontour/contour/pull/3033) does not include validation logic.
+Having a user-defined translation table makes adding a field like `content-id` require changes in two places (one to define `content-id` in the translation table and one to add it to the `json-fields` list).
+
+In this alternative, the example configuration would be written as the following:
+```yaml
+json-fields:
+   - @timestamp
+   - method
+   - content-id
+   - response_duration
+extra-json-fields:
+   content-id: %REQ(X-CONTENT-ID)%
+   response_duration: %RESPONSE_DURATION%
+```
+
+### Map Syntax
+The proposed implementation would require parsing each value to split at the first `=`.
+Instead of using an `=` within the string to separate a key and value, the Contour configuration could be a YAML key and value.
+In this approach, the example configuration would look like the following:
+```yaml
+json-fields:
+   @timestamp:
+   method:
+   content-id: %REQ(X-CONTENT-ID)%
+   response_duration: %RESPONSE_DURATION%
+```
+
+This approach is not backward-compatible because the structure of `json-fields` is changing from an array to a map.
+Existing configuration would not work.
+The other drawback of this approach is aesthetic: it could be argued that empty values for the keys with default translations is ugly.
+
+## Security Considerations
+Misconfiguration of the property can leak unwanted information to the access log file.
+
+## Compatibility
+Both original and the alternative translation table high-level designs don't introduce breaking changes, but the map syntax alternative would be a breaking change.
+The current implementation skips unknown fields so any new custom fields will be ignored.
+
+## Open Issues
+
+* Should all the fields with obvious translations consistently either require parameterization or not?
+  - In the current implementation, `method` becomes `"method": "%REQ(:METHOD)%"`.
+  - When the new implementation adds a way to achieve the JSON config `"response_duration": "%RESPONSE_DURATION%"`, should the user write:
+    * `response_duration`, or
+    * `response_durationn=%RESPONSE_DURATION%`
```

Comment:
The built-in mappings still need to be supported. I would not want to force people using the existing field mappings to manually remake their configuration.

Reply:
We agree that built-in mappings need to be supported. The question here is whether fields that could be added to those mappings should be.
* `method` will still work.
* should `response_duration` work?

If `response_duration` works, then the interface is consistent, but the implication is that the translation table will be used forever (despite users having a way to write `response_duration` without the translation table).
If `response_duration=%RESPONSE_DURATION%` is the only option for users, then there is inconsistency between the original set that are in the translations and the new ones that require more spelling out.

Reply:
The original intent was that we add simple translations as required to the translation table; we figured that people could just request anything they needed added. With that in mind, I'm okay with adding whatever simple translations to the translation table (like `response_duration`, where it's a straight mapping between field name and Envoy template string), as part of this implementation.

Obviously, this will be frozen at this point in time, but perhaps a comment saying that adding things to the standard translations is fine would make it clearer that we're open to adding more?